### PR TITLE
Add new ARKitScenes example

### PR DESCRIPTION
**Pull request by pablovela5620**
```
pip install -r examples/python/arkitscenes/requirements.txt
examples/python/arkitscenes/main.py
```

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)

https://user-images.githubusercontent.com/25287427/223909532-6ec9cebd-aaed-4b2c-94fc-299984e918eb.mov

Hey all, I'm attaching a python example for the [ARKitScenes dataset](https://github.com/apple/ARKitScenes). Currently, I've only included the trajectory and depth for low-res (256x192) data, but I plan on adding in 

- [x] High-Resolution Depth
- [x] .ply reconstruction from ARKit Depth
- [ ] ~~Faro Laser point cloud~~ (Turns out to be much too large to log | 160 million points)
- [x] 3d bounding boxes for objects
- [x] 2d bounding boxes reprojected into relevant view for both low and high res

a few points/questions I had:

1. I had to include a rotation of the extrinsic/intrinsics as some captures are in landscape orientation while others are in portrait, rather than being able to modify them using the `log_view_coordinate` I had to do it manually since there's no connection between rotating the 2d image/depth and the camera parameters
2. there's no clear way to give a logged depth image a colormap via the `log_depth_image` like there is for back projecting 

----EDIT---- 
I added in a few issues I found when trying to log the include .ply file #1570 and #1571 
Added in 3d bounding boxes, noticed that I would need to project the 3d keypoints into the 2d frame myself for them. Not sure if this is something that we want to include in this PR or wait until there's a way to do it via the rerun-sdk natively 

----EDIT 2---- 

Added code to reproject the oriented bounding boxes, added a ton of complexity that can be stripped out once #1581 is done

### CODE REVIEW

**Comment by nikolausWest** [comment]
> I had to include a rotation of the extrinsic/intrinsics as some captures are in landscape orientation while others are in portrait, rather than being able to modify them using the log_view_coordinate I had to do it manually since there's no connection between rotating the 2d image/depth and the camera parameters

What happens if you just set the view_coordinates correctly for the 3D view wrt landscape/portrait? Is the downside just that  the 2D view has the wrong orientation but everything matches up correctly otherwise?

> there's no clear way to give a logged depth image a colormap via the log_depth_image like there is for back projecting

Yeah we don't yet have support for setting colormaps to depth images (in 2D) yet. They are just colored with turbo as a quick fix but all of colormapping needs a retake.

**Comment by pablovela5620** [comment]
> What happens if you just set the view_coordinates correctly for the 3D view wrt landscape/portrait? Is the downside just that the 2D view has the wrong orientation but everything matches up correctly otherwise?

Yep exactly that, just means the 2d view is the wrong orientation

> Yeah we don't yet have support for setting colormaps to depth images (in 2D) yet. They are just colored with turbo as a quick fix but all of color mapping needs a retake.

Okay understood!

**Comment by nikolausWest** [comment]
> Yep exactly that, just means the 2d view is the wrong orientation

I actually don't think this is the worst thing in the world. I think it's fine to leave it like that in favour of simpler code, and just leave a `TODO(Pablovela): Fix orientation in 2D view once #1556 is closed.`

**Comment by pablovela5620** [comment]
Okay, I'll leave an example so you can see exactly what I'm talking about. I'll remove the orientation code and add TODO. 

https://user-images.githubusercontent.com/25287427/224437170-369b4788-aa38-4857-9141-5c69971d6525.mov

**Comment by emilk** [comment]
Very cool - I'm looking forwards to seeing this PR evolve!

**Comment by pablovela5620** [comment]
Here it is with the included bounding boxes (with mesh invisible/visible)


https://user-images.githubusercontent.com/25287427/224556285-1e8790be-44a1-4bef-96c4-3098b6166bba.mov

File: examples/python/arkitscenes/main.py
Comment (nikolausWest) [inline examples/python/arkitscenes/main.py:85]: This is a little overly complex. I think it makes more sense to directly convert the angle axis representation of the rotation to quaternion. Could for instance do it with a combination of `Rotation.from_rotvec` and `Rotation.as_quat` from [scipy.spatial.transform.Rotation](https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.transform.Rotation.html#scipy.spatial.transform.Rotation)

You also don't need to invert the transform here since the `rr.log_rigid3` lets you specify transforms as either `child_from_parent` and `parent_from_child`.

Comment (nikolausWest) [inline examples/python/arkitscenes/main.py:85]: This is a little overly complex. I think it makes more sense to directly convert the angle axis representation of the rotation to quaternion. Could for instance do it with a combination of `Rotation.from_rotvec` and `Rotation.as_quat` from [scipy.spatial.transform.Rotation](https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.transform.Rotation.html#scipy.spatial.transform.Rotation)

You also don't need to invert the transform here since the `rr.log_rigid3` lets you specify transforms as either `child_from_parent` and `parent_from_child`.

File: examples/python/arkitscenes/main.py
Comment (nikolausWest) [inline examples/python/arkitscenes/main.py:142]:
```suggestion
    mesh_ply = trimesh.load(str(ply_path))
    mesh_gltf_bin = trimesh.exchange.gltf.export_glb(ply, include_normals=True)
    rr.log_mesh_file("world/mesh", MeshFormat.GLB, mesh_gltf_bin)
```

(Note: you also need to `from rerun.log.file import MeshFormat`)

You could also log it with `log_mesh` but this way you'll automatically get vertex colors without changing the code when #1580 lands.

File: examples/python/arkitscenes/main.py
Comment (nikolausWest) [inline examples/python/arkitscenes/main.py:150]: Consider using `kornia_rs` to read images faster and avoid the bgr->rgb conversion

File: examples/python/arkitscenes/main.py
Comment (nikolausWest) [inline examples/python/arkitscenes/main.py:152]: Shouldn't frame_id already be a `str`?

File: examples/python/arkitscenes/main.py
Comment (nikolausWest) [inline examples/python/arkitscenes/main.py:171]: See earlier comment regarding converting to quaternion directly and using `child_from_parent` `parent_from_child`

File: examples/python/arkitscenes/main.py
Comment (nikolausWest) [inline examples/python/arkitscenes/main.py:44]: Rather than log these with `label`, it would be nice to use `class_id` + annotation context (see for instance the example `tracking_hf_opencv`)

File: examples/python/arkitscenes/main.py
Comment (nikolausWest) [inline examples/python/arkitscenes/main.py:110]: Navigation in the "world" view is a bit tedious without setting an "up direction". For instance: `rr.log_view_coordinates("world", up="+Y", timeless=True)` (I don't actually know what the correct up axis is here)

**Review by nikolausWest — CHANGES_REQUESTED** [review]
Great start!

I added some comment for improvements. Also, I think we should go ahead and add the high resolution depth as well.

Before merging we might want to wait for some other stuff to land but we might as well take it as far as we can first

File: examples/python/arkitscenes/main.py
Comment (pablovela5620) [reply examples/python/arkitscenes/main.py:152]: You're right! I thought it was of type Path

File: examples/python/arkitscenes/main.py
Comment (pablovela5620) [reply examples/python/arkitscenes/main.py:150]: Unfortunately, the kornia_rs package giving me a bunch of errors when trying to install ( on my M1 mac         `= note: ld: library not found for -lturbojpeg`, though its already installed). Given that the image loading isn't a bottleneck, I don't think its worth adding as a dep v.s. just doing it natively via opencv. Let me know your thoughts though

File: examples/python/arkitscenes/main.py
Comment (pablovela5620) [reply examples/python/arkitscenes/main.py:44]: So I tried looking at both the documentation on what is provided by arkitscenes and the tracking_hf_opencv example but it looks like only the `label` is provided to differentiate. There is this `object_id` but its the same for all of the different objects so I'm not totally sure what that's there for. For now what I did was remove the color so there could be more differentiation. I'm not fully sure if that's what you meant, but a bit more context would be helpful

**Comment by pablovela5620** [comment]
One question I had is regarding the linter. On my local machine, `just py-lint` passed but on the CI it did not. I'm guessing this is because of the version of python being used (CI is 3.8 v.s. my 3.10). Is this something that should be specified [here](https://github.com/rerun-io/rerun/blob/main/rerun_py/README.md)?

File: examples/python/arkitscenes/main.py
Comment (pablovela5620) [reply examples/python/arkitscenes/main.py:44]: I also wanted to ask if you'd like me to reproject the 3d bounding boxes into the 2d camera view (similar to the objectron dataset) or if its better to wait for #1581

Comment (nikolausWest) [reply examples/python/arkitscenes/main.py:45]: > So I tried looking at both the documentation on what is provided by arkitscenes and the tracking_hf_opencv example but it looks like only the label is provided to differentiate.

Alright, makes sense. I just went off of the images in the original repo and assumed there was some kind of label map. Let's stick with just labels.

> I also wanted to ask if you'd like me to reproject the 3d bounding boxes into the 2d camera view (similar to the objectron dataset) or if its better to wait for https://github.com/rerun-io/rerun/issues/1581. For now I just added a todo regarding this

I think it's going to be a couple weeks at least before we add that proper so I'd manually reproject them and add a todo to remove the manual code once automatic reprojection is fixed

**Comment by nikolausWest** [comment]
> One question I had is regarding the linter. On my local machine, just py-lint passed but on the CI it did not. I'm guessing this is because of the version of python being used (CI is 3.8 v.s. my 3.10). Is this something that should be specified [here](https://github.com/rerun-io/rerun/blob/main/rerun_py/README.md)?

Yeah, Rerun supports Python 3.8+ so CI needs to run with Python 3.8 to avoid features that depend on later versions creeping in. It used to say front and center but apparently no longer. The only place it's really written down is in the [pyproject.toml ](https://github.com/rerun-io/rerun/blob/f40752a993dc8e81b88e83fea7a03acc3ab7098d/rerun_py/pyproject.toml#L18) so it should probably say in the README as well

File: examples/python/arkitscenes/main.py
Comment (nikolausWest) [reply examples/python/arkitscenes/main.py:136]: Sounds good

**Comment by nikolausWest** [comment]
Btw, is it a lot of work to get the Faro laser scanner in as well?

File: examples/python/arkitscenes/download_dataset.py
Comment (nikolausWest) [inline examples/python/arkitscenes/download_dataset.py:1]:
```suggestion
# Copied from https://github.com/apple/ARKitScenes/blob/main/download_data.py
# Licensing information: https://github.com/apple/ARKitScenes/blob/main/LICENSE
```

File: examples/python/arkitscenes/main.py
Comment (nikolausWest) [inline examples/python/arkitscenes/main.py:134]:
```suggestion
        rr.set_time_sequence("frame_id", frame_id)
```

File: examples/python/arkitscenes/main.py
Comment (nikolausWest) [inline examples/python/arkitscenes/main.py:145]:
```suggestion
            camera_from_world = poses_from_traj[frame_id]
```

Comment (nikolausWest) [reply examples/python/arkitscenes/main.py:145]: Also, `poses_from_traj` should probably be called something like `poses_from_timestamp` or something (depending on what you decide to call frame_id)

File: examples/python/arkitscenes/main.py
Comment (nikolausWest) [inline examples/python/arkitscenes/main.py:63]:
```suggestion
def read_camera_from_world(traj_string: str) -> Tuple[str, Tuple[npt.NDArray[np.float64], npt.NDArray[np.float64]]]:
    """
    Reads out camera_from_world transform from trajectory string.

    Args:
        traj_string: A space-delimited file where each line represents a camera position at a particular timestamp.
            The file has seven columns:
            * Column 1: timestamp
            * Columns 2-4: rotation (axis-angle representation in radians)
            * Columns 5-7: translation (usually in meters)

    Returns
    -------
    timestamp: float
        timestamp in seconds
    camera_from_world: tuple of two numpy arrays
        A tuple containing a translation vector and a quaternion that represent the camera_from_world transform
```

File: examples/python/arkitscenes/main.py
Comment (pablovela5620) [reply examples/python/arkitscenes/main.py:134]: yea its a bit confusing, based on the original repo [here](https://github.com/apple/ARKitScenes/blob/e2e975128a0a9695ea56fa215fe76b4295241538/threedod/benchmark_scripts/utils/tenFpsDataLoader.py#L246) and [here](https://github.com/apple/ARKitScenes/blob/e2e975128a0a9695ea56fa215fe76b4295241538/threedod/benchmark_scripts/utils/tenFpsDataLoader.py#L257) you can see that the frame_id and timestaps  are rounded up to three decimal points. You can also see more info [here](https://github.com/apple/ARKitScenes/blob/main/threedod/README.md#data-organization-and-format-of-input-data) as to how the filenames are indexed by timestamps. I called it `frame_id` based on the original repo but I can rename it to something else. In the meantime I can add some more comments to try to clear up confusion

**Comment by pablovela5620** [comment]
> Btw, is it a lot of work to get the Faro laser scanner in as well?

That and the high res depth maps will be what I look into next. I almost feel like they should be separate though because of some of the timestamping problems I had previously found. The original repo splits things up into a `depth_upsampling` dataset and a `3d object detection` dataset. It may be worth doing something similar with the faro lidar scan and high res depth maps

File: examples/python/arkitscenes/main.py
Comment (nikolausWest) [reply examples/python/arkitscenes/main.py:147]: Makes sense. Yeah, commenting and a slightly more descriptive name would fix this for me.

**Comment by nikolausWest** [comment]
> That and the high res depth maps will be what I look into next. I almost feel like they should be separate though because of some of the timestamping problems I had previously found. The original repo splits things up into a depth_upsampling dataset and a 3d object detection dataset. It may be worth doing something similar with the faro lidar scan and high res depth maps

I'd like to understand those timestamping problems. It should be possible to handle well with Rerun I hope. You might have to use separate timelines though if the clocks aren't synched at all

**Comment by emilk** [comment]
Mesh vertex colors now work!

image

**Comment by emilk** [comment]
The depth images now look better too:

image

**Comment by pablovela5620** [comment]
Need to do a bit more cleanup but here's the current version

https://user-images.githubusercontent.com/25287427/227559993-eaa5370d-6959-4f9c-baf5-885dbd6a3447.mov

**Comment by nikolausWest** [comment]
> Need to do a bit more cleanup but here's the current version

Looking pretty cool!

File: examples/python/arkitscenes/main.py
Comment (nikolausWest) [inline examples/python/arkitscenes/main.py:88]:
```suggestion
        rot = R.from_matrix(rotation).inv()

        rr.log_obb(
            f"world/annotations/box-{uid}-{label}",
            half_size=scale,
            position=transform,
            rotation_q=rot.as_quat(),
            label=label,
            color=color_list[i],
            timeless=True,
        )
        
        box3d = compute_box_3d(scale, transform, rotation)
        bbox_list.append(box3d)
        bbox_labels.append(label)
```

I'd generally like to group everything that can be removed once #1581 is fixed together to make it easier to remove later and to make it easier to see what code _shouldn't_ be needed.

File: examples/python/arkitscenes/main.py
Comment (nikolausWest) [inline examples/python/arkitscenes/main.py:96]: Mention this can be removed once #1581 is fixed

File: examples/python/arkitscenes/main.py
Comment (nikolausWest) [inline examples/python/arkitscenes/main.py:133]: Mention this can be removed once #1581 is fixed

File: examples/python/arkitscenes/main.py
Comment (nikolausWest) [inline examples/python/arkitscenes/main.py:127]: This is really nice

File: examples/python/arkitscenes/main.py
Comment (nikolausWest) [inline examples/python/arkitscenes/main.py:177]: Mention this can be removed once #1581 is fixed

File: examples/python/arkitscenes/main.py
Comment (nikolausWest) [inline examples/python/arkitscenes/main.py:247]: Mention this can be removed once #1581 is fixed

File: examples/python/arkitscenes/main.py
Comment (nikolausWest) [inline examples/python/arkitscenes/main.py:63]: I wish you didn't have to do this. Very painful to see. Could you open an issue for us to add the functionality needed for you to not need this and add a todo here that refers to it?

If I'm not mistaken the issue is related to:
- You want synched colors in 2D and 3D
- You have labels, but no real label map (you don't know how many labels / classes there are and you don't have any class ids)

Is that correct?

I realize this specific instance will also not be needed once #1581 is fixed, but I could imagine the same problem arising in other circumstances and I'd like it if we could make it less annoying to deal with.

File: examples/python/arkitscenes/main.py
Comment (nikolausWest) [inline examples/python/arkitscenes/main.py:435]:
```suggestion
def main() -> None:
    parser = argparse.ArgumentParser(description="Visualizes the ARKitScenes dataset using the Rerun SDK.")
    parser.add_argument(
        "--video-id",
        type=str,
        choices=AVAILABLE_RECORDINGS,
        default=AVAILABLE_RECORDINGS[0],
        help="Video ID of the ARKitScenes Dataset",
    )
    rr.script_add_args(parser)
    args = parser.parse_args()

    rr.script_setup(args, "arkitscenes")
    recording_path = ensure_recording_available(args.video_id)
    log_arkit(recording_path)

    rr.script_teardown(args)
    
if __name__ == "__main__":
    main()
```

File: examples/python/arkitscenes/main.py
Comment (nikolausWest) [inline examples/python/arkitscenes/main.py:391]:
```suggestion
```

**Review by nikolausWest — CHANGES_REQUESTED** [review]
Some smaller fixes but once done I think this should be moved from draft -> "ready for review"

File: examples/python/arkitscenes/main.py
Comment (pablovela5620) [reply examples/python/arkitscenes/main.py:63]: Yep you have it right, I'll open up an issue for this

Comment (pablovela5620) [reply examples/python/arkitscenes/main.py:65]: Opened up an issue #1728

File: examples/python/arkitscenes/main.py
Comment (nikolausWest) [inline examples/python/arkitscenes/main.py:65]:
```suggestion
    # Generate a color per object that can be reused across both 3D obb and their 2D projections
    # TODO(pablovela5620): Once #1581 or #1728 is resolved this can be removed
    color_positions = np.linspace(0, 1, num_objects)
    colormap = plt.cm.get_cmap("viridis")
    color_array_float = [colormap(pos) for pos in color_positions]
    color_list = [(int(r * 255), int(g * 255), int(b * 255), int(a * 255)) for r, g, b, a in color_array_float]
```

**Review by nikolausWest — APPROVED** [review]
I have a last fix I'd like to get in but otherwise lgtm.

I'd love a review / look from someone else in the team as well though since I think I might have become blind by now.

**Comment by emilk** [comment]
The depth cloud rendering for the high-red depth point cloud is really slow 😬 not the fault of this PR obviously, but it makes me doubt whether it is good to have it on by default… maybe there is some low-hanging fruit there though.

**Comment by nikolausWest** [comment]
> The depth cloud rendering for the high-red depth point cloud is really slow 😬 not the fault of this PR obviously, but it makes me doubt whether it is good to have it on by default… maybe there is some low-hanging fruit there though.

Alright, maybe we put it behind a feature flag (command line option) until we've fixed the large depth image speed

File: examples/python/arkitscenes/main.py
Comment (emilk) [inline examples/python/arkitscenes/main.py:60]: `log_obb` should be able to take float colors too, though I notice the type signature for it is wrong 🤦

**Review by emilk — APPROVED** [review]
Very cool demo.

It reveals a lot of shortcomings in Rerun. I think we should merge this and then plan in a day where we try to fix as many as possible of the reported issues.

Oh, and also: this makes me wanna create an alternative first-person eye for the 3D view, as an alternative to the current orbital eye. The former is better for when you are inside the thing you are looking at (e.g. inside a room), while the latter is better when you are observing something from the outside.

**Pull request merged**
